apis/placement/v1alpha1/stagerollout_types.go

```diff
@@ -0,0 +1,299 @@
+type StageRollout struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of StageRollout.
+	// +required
+	Spec StageRolloutSpec `json:"spec"`
+
+	// The observed status of StageRollout.
+	// +optional
+	Status StageRolloutStatus `json:"status,omitempty"`
+}
+
+// StageRolloutSpec defines the desired the rollout sequence stage by stage.
+type StageRolloutSpec struct {
+	// The name of the CRP that this rollout is applied to.
+	// There can only be one active rollout for each ClusterResourcePlacement,
+	// we will fail the rollout if there is already an active rollout.
+	// +required
+	ParentCRPName string `json:"parentCRPName"`
+
+	// The resource snapshot index that this rollout is based on.
+	// The rollout will use the latest resource snapshot if this is empty.
+	// +optional
+	ResourceSnapShotIndex string `json:"resourceSnapShotIndex,omitempty"`
+
+	// Stage rollout configurations for each rollout stage.
+	// +required
+	Stages []StageRolloutConfig `json:"stages"`
+
+	// The common post rollout tasks that ALL stages needs to be completed before moving to the next stage.
+	// +optional
+	PostRolloutTasks PostRolloutTask `json:"postRolloutTasks,omitempty"`
+}
+
+// StageRolloutConfig describes a single rollout stage group configuration.
+type StageRolloutConfig struct {
+	// The name of the stage. This should be unique within the StageRollout.
+	// +required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected.
+	// We don't check if there are overlap between clusters selected by different stages.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The post rollout tasks that this stage needs to complete before moving to the next stage in-addition to
+	// the common post rollout tasks. This setting will override the common post rollout tasks if there are
+	// duplicate type of tasks with different settings.
+	// +optional
+	PostRolloutTasks PostRolloutTask `json:"postRolloutTasks,omitempty"`
+
+	// TODO: Add alerting configuration.
+
+}
+
+// PostRolloutTask is the collection of post rollout tasks that ALL need to be completed before we can move to the next stage.
+type PostRolloutTask struct {
+	// The name of the RolloutApprovalRequest we need to create to get the approval for the next stage.
+	// An RolloutApprovalRequest object will be created if this is not empty.
+	// +optional
+	ApprovalRequestName string `json:"approvalTask,omitempty"`
```

Comment:
keptn uses an generic data type to indicate the state and triggers the task by the state,
https://v1.keptn.sh/docs/1.0.x/define/triggers/


```
sequences:
  - name: "rollback"
    triggeredOn:
      - event: "production.delivery.finished"
        selector:
          match:
            result: failed
```

 https://github.com/keptn/spec/blob/0.2.2/cloudevents.md#evaluation-triggered

The pros is we don't need to create a CRD for each task.

Reply:
who will send the cloudEvent? How is it send and received?

Reply:
my bad, i thought it was a CR.